src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedCatchParameterShouldBeUnnamedCheck.java

```diff
@@ -0,0 +1,270 @@
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ * Ensures that catch parameters that are not used are declared as an unnamed variable.
+ * </p>
+ * <p>
+ * Rationale:
+ * </p>
+ * <ul>
+ *     <li>
+ *         Improves code readability by clearly indicating which parameters are unused.
+ *     </li>
+ *     <li>
+ *         Follows Java conventions for denoting unused parameters with an underscore ({@code _}).
+ *     </li>
+ * </ul>
+ * <p>
+ * See the <a href="https://docs.oracle.com/en/java/javase/21/docs/specs/unnamed-jls.html">
+ * Java Language Specification</a> for more information about unnamed variables.
+ * </p>
+ * <p>
+ * <b>Attention</b>: This check should be activated only on source code
+ * that is compiled by jdk21 or higher;
+ * unnamed catch parameters came out as the first preview in Java 21.
+ * </p>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code unused.catch.parameter}
+ * </li>
+ * </ul>
+ *
+ * @since 10.18.0
+ *
+ */
+
+@FileStatefulCheck
+public class UnusedCatchParameterShouldBeUnnamedCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_UNUSED_CATCH_PARAMETER = "unused.catch.parameter";
+
+    /**
+     * Invalid parents of the catch parameter identifier.
+     */
+    private static final int[] INVALID_CATCH_PARAM_IDENT_PARENTS = {
+        TokenTypes.DOT,
+        TokenTypes.LITERAL_NEW,
+        TokenTypes.METHOD_CALL,
+        TokenTypes.TYPE,
+    };
+
+    /**
+     * Keeps track of the catch parameters in a block.
+     */
+    private final Deque<CatchParameterDetails> catchParameters = new ArrayDeque<>();
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.LITERAL_CATCH,
+            TokenTypes.IDENT,
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        if (ast.getType() == TokenTypes.LITERAL_CATCH) {
+            final CatchParameterDetails catchParameter = new CatchParameterDetails(ast);
+            catchParameters.push(catchParameter);
+        }
+        else if (isCatchParameterIdentifierCandidate(ast) && !isLeftHandOfAssignment(ast)) {
+            // we do not count reassignment as usage
```

Comment:
The reassignment is never used in this case. What is the point in keeping it around?

Reply:
> The reassignment is never used in this case

How do we know unless we are tracking it also? When it talks about re-assignment, I thought we meant weird code like:
````
} catch (Exception e) {
  Except t = e;
  t.printStackTrace();
}
````

Also my example has the type, but I assume it can also be re-assigned a field outside the scope or a field.

Reply:
```
} catch (Exception e) {
  Except t = e;
  t.printStackTrace();
}
```
but this is not an LHS assignment. My code should not violate this example because `e` is actually used (another variable used its value)

```
} catch (Exception e) {
  e = t;  // e is a LHS value and it is not used
  t.printStackTrace();
}
```
ideally, we should remove the reassignment and make 'e' unamed.

Reply:
https://github.com/checkstyle/checkstyle/blob/827ccf0a7da00b12d3480027540569203c389555/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unusedcatchparametershouldbeunnamed/InputUnusedCatchParameterShouldBeUnnamed.java#L52-L60

Reply:
Sorry, I got the side of the assignment wrong. I see, it is still unused even though it was re-assigned.